core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -1595,6 +1590,7 @@ static Map<Integer, Integer> identityMap(int count) {
   Frame register(RelNode rel, RelNode newRel,
       Map<Integer, Integer> oldToNewOutputs,
       NavigableMap<CorDef, Integer> corDefOutputs) {
+    newRel = RelOptUtil.copyRelHints(rel, newRel);
```

Comment:
@rubenada I have one concern about using `RelOptUtil#copyRelHints`, if someday, we introduced another new `RelNode` and it's not `Hintable`, the hint will still be lost. E.g. rel is `Hintable`, and `newRel` is not `Hintable`.
Do you think that we need to do some propagation instead of coping?

Reply:
Uhm, good point. I guess it will be safer to use `RelOptUtil.propagateRelHints` instead of `RelOptUtil#copyRelHints` to cover also the cases that you mention.
@danny0405 IIRC you were the one that implemented the Hints mechanism, do you agree?

Reply:
Thanks for the fix @rubenada , the fix generally makes sense, event though there are redundant copy of hints, the fix avoids some buggy corner cases.

I was thinking, based on the fact that most of the kind of `RelNode` now are `Hintable`, we may need to start considering making the `AbstractRelNode` implement `Hintable` directly. And give some default implementations for some of the nodes.

Reply:
Thanks for the feedback @danny0405 , I agree.
The current patch was just a quick fix to avoid a regression in the imminent release. I have created a follow-up ticket (https://issues.apache.org/jira/browse/CALCITE-5188) where we can discuss more long-term solutions, like the one you propose.